File: auntieos-admin/android/app/src/test/java/com/tribetails/auntieos/ui/components/SlowWaitTest.kt
package com.tribetails.auntieos.ui.components

import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * The threshold decision as a decision table, with no Compose host.
 *
 * The fourth copy of one rule. Its twins are `auntieos-admin/src/lib/
 * slowWait.test.ts`, `mytribe/web/src/lib/slowWait.test.ts` and
 * `mytribe/src/commonTest/.../SlowWaitTest.kt`; between them they pin the same
 * number and the same offline behaviour on every client the operator can pick
 * up. If one drifts, exactly one of these four fails and says which.
 *
 * A plain JUnit test rather than a Compose one on purpose. `AuntieSpinner` runs
 * an infinite transition, so a composition containing it is never idle, and a
 * UI test would have to drive a virtual clock to a ten-second deadline to reach
 * the assertion. The rule is arithmetic; this is where arithmetic is tested.
 */
class SlowWaitTest {

    @Test
    fun `idle when nothing is in flight`() {
        assertEquals(WaitPhase.Idle, waitPhase(startedAtMs = null, nowMs = 1_000_000, online = true))
    }

    @Test
    fun `waiting the instant a wait starts`() {
        assertEquals(WaitPhase.Waiting, waitPhase(startedAtMs = 1_000L, nowMs = 1_000L, online = true))
    }

    /**
     * The measured cold start here is 7.9s to 9.4s -- the number issue #714 was
     * filed about. A healthy cold start must not be dressed up as a fault, or
     * the operator learns to tap through a wait that was always going to land.
     */
    @Test
    fun `stays waiting across the whole measured cold start range`() {
        for (elapsed in listOf(7_900L, 8_500L, 9_400L, 9_999L)) {
            assertEquals(
                "a $elapsed ms wait is a normal cold start and must not escalate",
                WaitPhase.Waiting,
                waitPhase(startedAtMs = 0L, nowMs = elapsed, online = true),
            )
        }
    }

    @Test
    fun `escalates exactly at the threshold`() {
        assertEquals(WaitPhase.Waiting, waitPhase(0L, SLOW_WAIT_MS - 1, online = true))
        assertEquals(WaitPhase.Slow, waitPhase(0L, SLOW_WAIT_MS, online = true))
    }

    /**
     * 10s leaves the other half of `fns.ts#CALLABLE_TIMEOUT_MS` (20s) for the
     * operator to read the offer and use it. An escalation offered later than
     * the budget would never be seen: the callable would already have rejected.
     */
    @Test
    fun `escalates with the callable budget still running`() {
        assertTrue(SLOW_WAIT_MS < 20_000L)
        assertEquals(WaitPhase.Slow, waitPhase(0L, 19_999L, online = true))
    }

    /**
     * THE DISCRIMINATING CASE. A device with no signal is not waiting on a slow
     * server and must never be handed a Sync button that fails the moment it is
     * pressed.
     */
    @Test
    fun `never escalates while offline however long the wait`() {
        assertEquals(WaitPhase.Waiting, waitPhase(0L, 60_000L, online = false))
        assertEquals(WaitPhase.Waiting, waitPhase(0L, 600_000L, online = false))
    }

    @Test
    fun `escalates on the remaining time when signal returns mid wait`() {
        assertEquals(WaitPhase.Slow, waitPhase(0L, 30_000L, online = true))
    }

    @Test
    fun `honours a caller supplied threshold`() {
        assertEquals(WaitPhase.Waiting, waitPhase(0L, 2_000L, online = true, thresholdMs = 3_000L))
        assertEquals(WaitPhase.Slow, waitPhase(0L, 3_000L, online = true, thresholdMs = 3_000L))
    }
}

File: auntieos-admin/src/lib/slowWait.test.ts
import { describe, expect, it } from 'vitest';
import { SLOW_WAIT_MS, waitPhase } from './slowWait';

/**
 * The threshold decision, exercised as a decision table with no React, no
 * jsdom and no fake timers -- the same way `lib/async.ts` and the portal's
 * `lib/queryState.ts` are tested, and for the same reason: this is the rule the
 * operator's ruling turns into, and a rule that can only be reached by
 * rendering a component and pushing timers gets tested once and then trusted.
 */
describe('waitPhase', () => {
  const online = true;

  it('is idle when nothing is in flight', () => {
    expect(waitPhase({ startedAt: null, now: 1_000_000, online })).toBe('idle');
  });

  it('is waiting the instant a wait starts', () => {
    expect(waitPhase({ startedAt: 1_000_000, now: 1_000_000, online })).toBe('waiting');
  });

  /**
   * The measured cold start on this project is 7.9s to 9.4s. A healthy cold
   * start must NOT be dressed up as a fault, or the operator learns to tap
   * through a wait that was always going to land on its own.
   */
  it('stays waiting across the whole measured cold-start range', () => {
    for (const elapsed of [7_900, 8_500, 9_400, 9_999]) {
      expect(waitPhase({ startedAt: 0, now: elapsed, online })).toBe('waiting');
    }
  });

  it('escalates exactly at the threshold', () => {
    expect(waitPhase({ startedAt: 0, now: SLOW_WAIT_MS - 1, online })).toBe('waiting');
    expect(waitPhase({ startedAt: 0, now: SLOW_WAIT_MS, online })).toBe('slow');
  });

  /**
   * 10s leaves the other half of `lib/fns.ts#CALLABLE_TIMEOUT_MS` (20s) for the
   * operator to read the affordance and use it. An escalation offered later
   * than the budget would never be seen, because the callable would already
   * have rejected into the error path.
   */
  it('escalates with the callable budget still running', () => {
    const CALLABLE_TIMEOUT_MS = 20_000;
    expect(SLOW_WAIT_MS).toBeLessThan(CALLABLE_TIMEOUT_MS);
    expect(waitPhase({ startedAt: 0, now: CALLABLE_TIMEOUT_MS - 1, online })).toBe('slow');
  });

  /**
   * THE DISCRIMINATING CASE. A device with no signal is not waiting on a slow
   * server, and must never be handed a Sync button that cannot work. Offline
   * keeps its own treatment (PR #805 in the portal, RouteError here).
   */
  it('never escalates while offline, however long the wait', () => {
    expect(waitPhase({ startedAt: 0, now: 60_000, online: false })).toBe('waiting');
    expect(waitPhase({ startedAt: 0, now: 600_000, online: false })).toBe('waiting');
  });

  it('escalates on the remaining time when signal returns mid-wait', () => {
    // Started offline, 30s ago; the connection is back, so the wait is now a
    // slow server and is allowed to say so.
    expect(waitPhase({ startedAt: 0, now: 30_000, online: true })).toBe('slow');
  });

  it('honours a caller-supplied threshold', () => {
    expect(waitPhase({ startedAt: 0, now: 2_000, online, thresholdMs: 3_000 })).toBe('waiting');
    expect(waitPhase({ startedAt: 0, now: 3_000, online, thresholdMs: 3_000 })).toBe('slow');
  });
});
